src/credentials/FabricTable.cpp

```diff
@@ -1248,97 +1383,548 @@ CHIP_ERROR FabricTable::AllocatePendingOperationalKey(Optional<FabricIndex> fabr
+        Credentials::P256PublicKeySpan publicKeySpan;
+        ReturnErrorOnFailure(FetchRootCert(pendingFabricIndex, tempRcacSpan));
+        ReturnErrorOnFailure(ExtractPublicKeyFromChipCert(tempRcacSpan, publicKeySpan));
+        candidateRootKey = publicKeySpan;
+    }
+
+    for (auto & existingFabric : *this)
+    {
+        if (existingFabric.GetFabricId() == fabricId)
+        {
+            P256PublicKey existingRootKey;
+            ReturnErrorOnFailure(FetchRootPubkey(existingFabric.GetFabricIndex(), existingRootKey));
+
+            if (existingRootKey.Matches(candidateRootKey))
+            {
+                outMatchingFabricIndex = existingFabric.GetFabricIndex();
+                return CHIP_NO_ERROR;
+            }
+        }
+    }
+
+    // Did not find: set outMatchingFabricIndex to kUndefinedFabricIndex
+    outMatchingFabricIndex = kUndefinedFabricIndex;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR FabricTable::AddNewPendingFabricCommon(const ByteSpan & noc, const ByteSpan & icac, uint16_t vendorId,
+                                                  Crypto::P256Keypair * existingOpKey, bool isExistingOpKeyExternallyOwned,
+                                                  FabricIndex * outNewFabricIndex)
+{
+    VerifyOrReturnError(mOpCertStore != nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(outNewFabricIndex != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    static_assert(kMaxValidFabricIndex <= UINT8_MAX, "Cannot create more fabrics than UINT8_MAX");
+
+    // We should already have a pending root when we get here
+    VerifyOrReturnError(mStateFlags.Has(StateFlags::kIsTrustedRootPending), CHIP_ERROR_INCORRECT_STATE);
+    // We should not have pending update when we get here
+    VerifyOrReturnError(!mStateFlags.Has(StateFlags::kIsUpdatePending), CHIP_ERROR_INCORRECT_STATE);
+
+    EnsureNextAvailableFabricIndexUpdated();
+    FabricIndex fabricIndexToUse = kUndefinedFabricIndex;
+    if (mNextAvailableFabricIndex.HasValue())
+    {
+        // Mark we have some pending data for a given fabric.
+        fabricIndexToUse = mNextAvailableFabricIndex.Value();
+    }
+    else
+    {
+        // Fabric addition, but adding fabric would fail on table full: let's not allocate a fabric
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    if (existingOpKey == nullptr)
+    {
+        // If existing operational key not provided, we need to have a keystore present.
+        // It should already have an operational key pending.
+        VerifyOrReturnError(mOperationalKeystore != nullptr, CHIP_ERROR_KEY_NOT_FOUND);
+        // Make sure we have an operational key, pending or not
+        VerifyOrReturnError(mOperationalKeystore->HasOpKeypairForFabric(fabricIndexToUse) ||
+                                mOperationalKeystore->HasPendingOpKeypair(),
+                            CHIP_ERROR_KEY_NOT_FOUND);
+    }
+
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndexToUse), CHIP_ERROR_INVALID_FABRIC_INDEX);
```

Comment:
How could this happen?  We'd need to have an invalid `mNextAvailableFabricIndex`, right?  Shouldn't we verify that doesn't happen in the one place where it could, where we de-serialize it?

If we do keep this check, please move it to before the opkey block, where we use `fabricIndexToUse`.

Reply:
Not sure how to do so, will add a comment for now